### ISSUE TITLE
certbot-apache cannot find the SSL module on openSUSE

### ISSUE THREAD

**Issue by Quembar** [certbot/certbot#8711]
**## My operating system is (include version):**
openSUSE Leap 15.2

**## I installed Certbot with (certbot-auto, OS package manager, pip, etc):**
OS package manager

**## I ran this command and it produced this output:**
Get this message when renew:
Could not find ssl_module; not disabling session tickets.

**## Certbot's behavior differed from what I expected because:**
renew the certs as usual

**## Here is a Certbot log showing the issue (if available):**
2021-03-11 10:09:07,946:DEBUG:certbot._internal.main:certbot version: 1.4.0
2021-03-11 10:09:07,948:DEBUG:certbot._internal.main:Arguments: []
2021-03-11 10:09:07,948:DEBUG:certbot._internal.main:Discovered plugins: PluginsRegistry(PluginEntryPoint#apache,PluginEntryPoint#manual,PluginEntryPoint#null,PluginEntryPoint#standalone,PluginEntryPoint#webroot)
2021-03-11 10:09:07,967:DEBUG:certbot._internal.log:Root logging level set at 20
2021-03-11 10:09:07,968:INFO:certbot._internal.log:Saving debug log to /var/log/letsencrypt/letsencrypt.log
2021-03-11 10:09:07,969:DEBUG:certbot._internal.plugins.selection:Requested authenticator None and installer None
2021-03-11 10:09:08,052:DEBUG:certbot_apache._internal.configurator:Apache version is 2.4.43
2021-03-11 10:09:08,596:WARNING:certbot_apache._internal.configurator:Could not find ssl_module; not disabling session tickets.
2021-03-11 10:09:08,597:DEBUG:certbot._internal.plugins.selection:Single candidate plugin: * apache
Description: Apache Web Server plugin
Interfaces: IAuthenticator, IInstaller, IPlugin
Entry point: apache = certbot_apache._internal.entrypoint:ENTRYPOINT
Initialized: <certbot_apache._internal.override_suse.OpenSUSEConfigurator object at 0x7f99d09b19b0>
Prep: True
2021-03-11 10:09:08,599:DEBUG:certbot._internal.plugins.selection:Selected authenticator <certbot_apache._internal.override_suse.OpenSUSEConfigurator object at 0x7f99d09b19b0> and installer <certbot_apache._internal.override_suse.OpenSUSEConfigurator object at 0x7f99d09b19b0>
2021-03-11 10:09:08,599:INFO:certbot._internal.plugins.selection:Plugins selected: Authenticator apache, Installer apache
2021-03-11 10:09:08,604:DEBUG:certbot._internal.main:Picked account: <Account(RegistrationResource(body=Registration(key=None, contact=(), agreement=None, status=None, terms_of_service_agreed=None, only_return_existing=None, external_account_binding=None), uri='https://acme-v02.api.letsencrypt.org/acme/acct/81739808', new_authzr_uri=None, terms_of_service=None), e06dc68747fca99f419acc1399b45bd5, Meta(creation_dt=datetime.datetime(2020, 3, 27, 11, 27, 25, tzinfo=<UTC>), creation_host='kkhgcrproxy01.intern.rvkgc.de'))>
2021-03-11 10:09:08,605:DEBUG:acme.client:Sending GET request to https://acme-v02.api.letsencrypt.org/directory.
2021-03-11 10:09:08,607:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): acme-v02.api.letsencrypt.org:443
2021-03-11 10:09:09,251:DEBUG:urllib3.connectionpool:https://acme-v02.api.letsencrypt.org:443 "GET /directory HTTP/1.1" 200 658
2021-03-11 10:09:09,252:DEBUG:acme.client:Received response:
HTTP 200
Server: nginx
Date: Thu, 11 Mar 2021 09:09:09 GMT
Content-Type: application/json
Content-Length: 658
Connection: keep-alive
Cache-Control: public, max-age=0, no-cache
X-Frame-Options: DENY
Strict-Transport-Security: max-age=604800

{
  "B_6cPdKdINk": "https://community.letsencrypt.org/t/adding-random-entries-to-the-directory/33417",
  "keyChange": "https://acme-v02.api.letsencrypt.org/acme/key-change",
  "meta": {
    "caaIdentities": [
      "letsencrypt.org"
    ],
    "termsOfService": "https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf",
    "website": "https://letsencrypt.org"
  },
  "newAccount": "https://acme-v02.api.letsencrypt.org/acme/new-acct",
  "newNonce": "https://acme-v02.api.letsencrypt.org/acme/new-nonce",
  "newOrder": "https://acme-v02.api.letsencrypt.org/acme/new-order",
  "revokeCert": "https://acme-v02.api.letsencrypt.org/acme/revoke-cert"
}
2021-03-11 10:09:12,278:DEBUG:certbot._internal.log:Exiting abnormally:
Traceback (most recent call last):
  File "/usr/bin/certbot", line 11, in <module>
    load_entry_point('certbot==1.4.0', 'console_scripts', 'certbot')()
  File "/usr/lib/python3.6/site-packages/certbot/main.py", line 15, in main
    return internal_main.main(cli_args)
  File "/usr/lib/python3.6/site-packages/certbot/_internal/main.py", line 1347, in main
    return config.func(config, plugins)
  File "/usr/lib/python3.6/site-packages/certbot/_internal/main.py", line 1095, in run
    domains, certname = _find_domains_or_certname(config, installer)
  File "/usr/lib/python3.6/site-packages/certbot/_internal/main.py", line 422, in _find_domains_or_certname
    raise errors.Error("Please specify --domains, or --installer that "
certbot.errors.Error: Please specify --domains, or --installer that will help in domain names autodiscovery, or --cert-name for an existing certificate name.

**Comment by Quembar** [comment]
BTW: # a2enmod -l
actions alias auth_basic authn_core authn_file authz_host authz_groupfile authz_core authz_user autoindex cgi dir env expires include log_config mime negotiation setenvif **ssl** socache_shmcb reqtimeout userdir rewrite proxy proxy_http proxy_connect

The module in question is loaded!

**Comment by bmw** [comment]
There are multiple things going on here.

The message `Could not find ssl_module; not disabling session tickets` is a bug in Certbot on OpenSUSE Leap 15.2 that I reproduced using our `master` branch as of writing this (although the error message changes to `ssl_module is statically linked but --apache-bin is missing; not disabling session tickets.` after upgrading Certbot and providing `--apache-bin` changes the error message to `Could not find OpenSSL version; not disabling session tickets.`.) The discussion in https://github.com/certbot/certbot/issues/8072 seems related to this. These messages are not fatal though and are unrelated to your renewal problem.

Looking at your logfile, I noticed:

1. You're using `certbot run` to renew and I'd recommend using `certbot renew`.
2. Using `certbot run`, I believe Certbot would have prompted you about the domains you want to obtain a certificate for and not answering that question would have resulted in this crash.

I'll keep this issue open to track fixing finding the SSL module on OpenSUSE and if you need additional help in handling the renewal problem, I recommend creating a thread at https://community.letsencrypt.org where there is a much larger community of people familiar with using Certbot who should be able to help you.

**Comment by bmw** [comment]
There are multiple things going on here.

The message `Could not find ssl_module; not disabling session tickets` is a bug in Certbot on OpenSUSE Leap 15.2 that I reproduced using our `master` branch as of writing this (although the error message changes to `ssl_module is statically linked but --apache-bin is missing; not disabling session tickets.` after upgrading Certbot and providing `--apache-bin` changes the error message to `Could not find OpenSSL version; not disabling session tickets.`.) The discussion in https://github.com/certbot/certbot/issues/8072 seems related to this. These messages are not fatal though and are unrelated to your renewal problem.

Looking at your logfile, I noticed:

1. You're using `certbot run` to renew and I'd recommend using `certbot renew`.
2. Using `certbot run`, I believe Certbot would have prompted you about the domains you want to obtain a certificate for and not answering that question would have resulted in this crash.

I'll keep this issue open to track fixing finding the SSL module on OpenSUSE and if you need additional help in handling the renewal problem, I recommend creating a thread at https://community.letsencrypt.org where there is a much larger community of people familiar with using Certbot who should be able to help you.

**Comment by deveric06** [comment]
Hi, I have the same problem. OpensSuse 15.2. Apache version is 2.4.43 - certbot 1.4.0 . ssl is enabled. and the message : Could not find ssl_module; not disabling session tickets. Thanks for any help

**Comment by Exzellius** [comment]
Hi,
the issue still exists for me on openSUSE Leap 15.3

**Comment by drbeco** [comment]
Hello. Just to share the way I solved this problem in Slackware 15.0.

I was getting the error:

`Unable to read ssl_module file; not disabling session tickets.`

Using `lsof` and `strace` I tracked down the problem. In my system, this module is located at:
  
`/usr/lib64/httpd/modules/mod_ssl.so`

but certbot was trying to open it at:

`/etc/httpd/lib64/httpd/modules`

I solved by creating a soft link:

```
mkdir -p /etc/httpd/lib64/httpd/modules
ln -s /usr/lib64/httpd/modules/mod_ssl.so /etc/httpd/lib64/httpd/modules
```

All good here now.
